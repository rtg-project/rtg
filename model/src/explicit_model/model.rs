use super::entity::{ExplicitEntity, ImplicitEntity};
use partial_struct::PartialStruct;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::rc::Rc;

#[derive(PartialStruct, Serialize, Deserialize, Debug, JsonSchema)]
#[serde(tag = "version", rename_all = "camelCase")]
#[partial(name = "ImplicitModel")]
#[partial_attribute(derive(Serialize, Deserialize, Debug, JsonSchema))]
#[partial_attribute(serde(tag = "version", rename_all = "camelCase"))]
pub enum ExplicitModel {
  #[serde(rename_all = "camelCase")]
  V1 {
    #[partial(nested_type = "Vec<Rc<ImplicitEntity>>")]
    entities: Vec<Rc<ExplicitEntity>>,
  },
}

// Tests
#[cfg(test)]
mod tests {
  use super::*;
  use crate::explicit_model::field::ExplicitField;
  use crate::explicit_model::sql_type;
  use insta::assert_json_snapshot;

  #[test]
  fn serialize() {
    let value = ExplicitModel::V1 {
      entities: vec![Rc::new(ExplicitEntity::DatabaseTable {
        name: "person".to_owned(),
        sql_schema_name: "public".to_owned(),
        sql_table_name: "person_table".to_owned(),
        graphql_entity_type_name: "Person".to_owned(),
        graphql_filter_type_name: "PersonWhereFilter".to_owned(),
        graphql_get_single_operation_name: "person".to_owned(),
        graphql_get_list_operation_name: "persons".to_owned(),
        graphql_get_connection_operation_name: "personConnection".to_owned(),
        graphql_default_order_by: "id_ASC".to_owned(),
        graphql_default_first: 10,
        graphql_default_offset: 0,
        fields: vec![
          Rc::new(ExplicitField::ScalarDatabaseColumn {
            name: "id".to_owned(),
            nullable: false,
            sql_type: sql_type::Type::Text,
            sql_column_name: "id_col".to_owned(),
            graphql_enabled: true,
            graphql_field_name: "id".to_owned(),
            graphql_type_name: "String".to_owned(),
            graphql_order_by_asc: "id_ASC".to_owned(),
            graphql_order_by_desc: "id_DESC".to_owned(),
          }),
          Rc::new(ExplicitField::ScalarDatabaseColumn {
            name: "drone".to_owned(),
            nullable: false,
            sql_type: sql_type::Type::Text,
            sql_column_name: "drone_col".to_owned(),
            graphql_enabled: true,
            graphql_field_name: "drone".to_owned(),
            graphql_type_name: "String".to_owned(),
            graphql_order_by_asc: "drone_ASC".to_owned(),
            graphql_order_by_desc: "drone_DESC".to_owned(),
          }),
        ],
      })],
    };

    assert_json_snapshot!(serde_json::to_string_pretty(&value).unwrap());
  }
}
