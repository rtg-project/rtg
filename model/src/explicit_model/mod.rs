pub mod entity;
pub mod field;
pub mod json_schema;
pub mod model;
pub mod sql_type;

// Tests
#[cfg(test)]
mod tests {

  use similar_asserts::assert_eq;
  #[test]
  fn it_works() {
    assert_eq!(2 + 2, 4);
  }

  #[test]
  fn it_works2() {
    assert_eq!(3 + 2, 5);
  }
}
