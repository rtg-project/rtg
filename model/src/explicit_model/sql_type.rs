// See rust-postgres
use derive_more::Display;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

// See https://github.com/sfackler/rust-postgres/blob/fc10985f9fdf0903893109bc951fb5891539bf97/postgres-types/src/type_gen.rs
#[derive(Serialize, Deserialize, Debug, PartialEq, JsonSchema, Clone, Display)]
#[serde(rename_all = "camelCase")]
pub enum Type {
  #[display(fmt = "bool")]
  Bool,
  #[display(fmt = "bytea")]
  Bytea,
  #[display(fmt = "char")]
  Char,
  #[display(fmt = "name")]
  Name,
  #[display(fmt = "int8")]
  Int8,
  #[display(fmt = "int2")]
  Int2,
  #[display(fmt = "int2vector")]
  Int2Vector,
  #[display(fmt = "int4")]
  Int4,
  #[display(fmt = "regproc")]
  Regproc,
  #[display(fmt = "text")]
  Text,
  #[display(fmt = "oid")]
  Oid,
  #[display(fmt = "tid")]
  Tid,
  #[display(fmt = "xid")]
  Xid,
  #[display(fmt = "cid")]
  Cid,
  #[display(fmt = "oidvector")]
  OidVector,
  #[display(fmt = "pgddlcommand")]
  PgDdlCommand,
  #[display(fmt = "json")]
  Json,
  #[display(fmt = "xml")]
  Xml,
  #[display(fmt = "xml[]")]
  XmlArray,
  #[display(fmt = "pgnodetree")]
  PgNodeTree,
  #[display(fmt = "json[]")]
  JsonArray,
  #[display(fmt = "tableamhandler")]
  TableAmHandler,
  #[display(fmt = "xid8[]")]
  Xid8Array,
  #[display(fmt = "indexamhandler")]
  IndexAmHandler,
  #[display(fmt = "point")]
  Point,
  #[display(fmt = "lseg")]
  Lseg,
  #[display(fmt = "path")]
  Path,
  #[display(fmt = "box")]
  Box,
  #[display(fmt = "polygon")]
  Polygon,
  #[display(fmt = "line")]
  Line,
  #[display(fmt = "line[]")]
  LineArray,
  #[display(fmt = "cidr")]
  Cidr,
  #[display(fmt = "cidr[]")]
  CidrArray,
  #[display(fmt = "float4")]
  Float4,
  #[display(fmt = "float8")]
  Float8,
  #[display(fmt = "unknown")]
  Unknown,
  #[display(fmt = "circle")]
  Circle,
  #[display(fmt = "circle[]")]
  CircleArray,
  #[display(fmt = "macaddr8")]
  Macaddr8,
  #[display(fmt = "macaddr8[]")]
  Macaddr8Array,
  #[display(fmt = "money")]
  Money,
  #[display(fmt = "money[]")]
  MoneyArray,
  #[display(fmt = "macaddr")]
  Macaddr,
  #[display(fmt = "inet")]
  Inet,
  #[display(fmt = "bool[]")]
  BoolArray,
  #[display(fmt = "bytea[]")]
  ByteaArray,
  #[display(fmt = "char[]")]
  CharArray,
  #[display(fmt = "name[]")]
  NameArray,
  #[display(fmt = "int2[]")]
  Int2Array,
  #[display(fmt = "int2vector[]")]
  Int2VectorArray,
  #[display(fmt = "int4[]")]
  Int4Array,
  #[display(fmt = "regproc[]")]
  RegprocArray,
  #[display(fmt = "text[]")]
  TextArray,
  #[display(fmt = "tid[]")]
  TidArray,
  #[display(fmt = "xid[]")]
  XidArray,
  #[display(fmt = "cid[]")]
  CidArray,
  #[display(fmt = "oidvector[]")]
  OidVectorArray,
  #[display(fmt = "bpchar[]")]
  BpcharArray,
  #[display(fmt = "varchar[]")]
  VarcharArray,
  #[display(fmt = "int8[]")]
  Int8Array,
  #[display(fmt = "point[]")]
  PointArray,
  #[display(fmt = "lseg[]")]
  LsegArray,
  #[display(fmt = "path[]")]
  PathArray,
  #[display(fmt = "box[]")]
  BoxArray,
  #[display(fmt = "float4[]")]
  Float4Array,
  #[display(fmt = "float8[]")]
  Float8Array,
  #[display(fmt = "polygon[]")]
  PolygonArray,
  #[display(fmt = "oid[]")]
  OidArray,
  #[display(fmt = "aclitem")]
  Aclitem,
  #[display(fmt = "aclitem[]")]
  AclitemArray,
  #[display(fmt = "macaddr[]")]
  MacaddrArray,
  #[display(fmt = "inet[]")]
  InetArray,
  #[display(fmt = "bpchar")]
  Bpchar,
  #[display(fmt = "varchar")]
  Varchar,
  #[display(fmt = "date")]
  Date,
  #[display(fmt = "time")]
  Time,
  #[display(fmt = "timestamp")]
  Timestamp,
  #[display(fmt = "timestamp[]")]
  TimestampArray,
  #[display(fmt = "date[]")]
  DateArray,
  #[display(fmt = "time[]")]
  TimeArray,
  #[display(fmt = "timestamptz")]
  Timestamptz,
  #[display(fmt = "timestamptz[]")]
  TimestamptzArray,
  #[display(fmt = "interval")]
  Interval,
  #[display(fmt = "interval[]")]
  IntervalArray,
  #[display(fmt = "numeric[]")]
  NumericArray,
  #[display(fmt = "cstring[]")]
  CstringArray,
  #[display(fmt = "timetz")]
  Timetz,
  #[display(fmt = "timetz[]")]
  TimetzArray,
  #[display(fmt = "bit")]
  Bit,
  #[display(fmt = "bit[]")]
  BitArray,
  #[display(fmt = "varbit")]
  Varbit,
  #[display(fmt = "varbit[]")]
  VarbitArray,
  #[display(fmt = "numeric")]
  Numeric,
  #[display(fmt = "refcursor")]
  Refcursor,
  #[display(fmt = "refcursor[]")]
  RefcursorArray,
  #[display(fmt = "regprocedure")]
  Regprocedure,
  #[display(fmt = "regoper")]
  Regoper,
  #[display(fmt = "regoperator")]
  Regoperator,
  #[display(fmt = "regclass")]
  Regclass,
  #[display(fmt = "regtype")]
  Regtype,
  #[display(fmt = "regprocedure[]")]
  RegprocedureArray,
  #[display(fmt = "regoper[]")]
  RegoperArray,
  #[display(fmt = "regoperator[]")]
  RegoperatorArray,
  #[display(fmt = "regclass[]")]
  RegclassArray,
  #[display(fmt = "regtype[]")]
  RegtypeArray,
  #[display(fmt = "record")]
  Record,
  #[display(fmt = "cstring")]
  Cstring,
  #[display(fmt = "any")]
  Any,
  #[display(fmt = "anyarray")]
  Anyarray,
  #[display(fmt = "void")]
  Void,
  #[display(fmt = "trigger")]
  Trigger,
  #[display(fmt = "languagehandler")]
  LanguageHandler,
  #[display(fmt = "internal")]
  Internal,
  #[display(fmt = "anyelement")]
  Anyelement,
  #[display(fmt = "record[]")]
  RecordArray,
  #[display(fmt = "anynonarray")]
  Anynonarray,
  #[display(fmt = "txidsnapshot[]")]
  TxidSnapshotArray,
  #[display(fmt = "uuid")]
  Uuid,
  #[display(fmt = "uuid[]")]
  UuidArray,
  #[display(fmt = "txidsnapshot")]
  TxidSnapshot,
  #[display(fmt = "fdwhandler")]
  FdwHandler,
  #[display(fmt = "pglsn")]
  PgLsn,
  #[display(fmt = "pglsn[]")]
  PgLsnArray,
  #[display(fmt = "tsmhandler")]
  TsmHandler,
  #[display(fmt = "pgndistinct")]
  PgNdistinct,
  #[display(fmt = "pgdependencies")]
  PgDependencies,
  #[display(fmt = "anyenum")]
  Anyenum,
  #[display(fmt = "tsvector")]
  TsVector,
  #[display(fmt = "tsquery")]
  Tsquery,
  #[display(fmt = "gtsvector")]
  GtsVector,
  #[display(fmt = "tsvector[]")]
  TsVectorArray,
  #[display(fmt = "gtsvector[]")]
  GtsVectorArray,
  #[display(fmt = "tsquery[]")]
  TsqueryArray,
  #[display(fmt = "regconfig")]
  Regconfig,
  #[display(fmt = "regconfig[]")]
  RegconfigArray,
  #[display(fmt = "regdictionary")]
  Regdictionary,
  #[display(fmt = "regdictionary[]")]
  RegdictionaryArray,
  #[display(fmt = "jsonb")]
  Jsonb,
  #[display(fmt = "jsonb[]")]
  JsonbArray,
  #[display(fmt = "anyrange")]
  AnyRange,
  #[display(fmt = "eventtrigger")]
  EventTrigger,
  #[display(fmt = "int4range")]
  Int4Range,
  #[display(fmt = "int4range[]")]
  Int4RangeArray,
  #[display(fmt = "numrange")]
  NumRange,
  #[display(fmt = "numrange[]")]
  NumRangeArray,
  #[display(fmt = "tsrange")]
  TsRange,
  #[display(fmt = "tsrange[]")]
  TsRangeArray,
  #[display(fmt = "tstzrange")]
  TstzRange,
  #[display(fmt = "tstzrange[]")]
  TstzRangeArray,
  #[display(fmt = "daterange")]
  DateRange,
  #[display(fmt = "daterange[]")]
  DateRangeArray,
  #[display(fmt = "int8range")]
  Int8Range,
  #[display(fmt = "int8range[]")]
  Int8RangeArray,
  #[display(fmt = "jsonpath")]
  Jsonpath,
  #[display(fmt = "jsonpath[]")]
  JsonpathArray,
  #[display(fmt = "regnamespace")]
  Regnamespace,
  #[display(fmt = "regnamespace[]")]
  RegnamespaceArray,
  #[display(fmt = "regrole")]
  Regrole,
  #[display(fmt = "regrole[]")]
  RegroleArray,
  #[display(fmt = "regcollation")]
  Regcollation,
  #[display(fmt = "regcollation[]")]
  RegcollationArray,
  #[display(fmt = "pgmcvlist")]
  PgMcvList,
  #[display(fmt = "pgsnapshot")]
  PgSnapshot,
  #[display(fmt = "pgsnapshot[]")]
  PgSnapshotArray,
  #[display(fmt = "xid8")]
  Xid8,
  #[display(fmt = "anycompatible")]
  Anycompatible,
  #[display(fmt = "anycompatiblearray")]
  Anycompatiblearray,
  #[display(fmt = "anycompatiblenonarray")]
  Anycompatiblenonarray,
  #[display(fmt = "anycompatiblerange")]
  AnycompatibleRange,
  #[serde(rename_all = "camelCase")]
  #[display(fmt = "{}", "_0")]
  Other(String),
}

#[cfg(test)]
mod tests {
  use super::*;
  use similar_asserts::assert_eq;

  #[test]
  fn serialize1() {
    let value = Type::Bit;

    match serde_json::to_string_pretty(&value) {
      Ok(string) => {
        assert_eq!(string, r#""bit""#);
      }
      Err(e) => panic!("{}", e),
    }
  }

  #[test]
  fn serialize2() {
    let value = Type::Int8;

    match serde_json::to_string_pretty(&value) {
      Ok(string) => {
        assert_eq!(string, r#""int8""#);
      }
      Err(e) => panic!("{}", e),
    }
  }

  #[test]
  fn serialize_other() {
    let value = Type::Other("wow".to_owned());

    match serde_json::to_string_pretty(&value) {
      Ok(string) => {
        assert_eq!(
          string,
          r#"{
  "other": "wow"
}"#
        );
      }
      Err(e) => panic!("{}", e),
    }
  }
}
