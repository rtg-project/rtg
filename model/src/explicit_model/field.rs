use super::sql_type;
use partial_struct::PartialStruct;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

#[derive(PartialStruct, Serialize, Deserialize, Debug, JsonSchema)]
#[serde(tag = "type", rename_all = "camelCase")]
#[partial(name = "ImplicitField")]
#[partial_attribute(derive(Serialize, Deserialize, Debug, JsonSchema))]
#[partial_attribute(serde(tag = "type", rename_all = "camelCase"))]
pub enum ExplicitField {
  #[serde(rename_all = "camelCase")]
  ScalarDatabaseColumn {
    name: String,
    nullable: bool,
    // #[serde(flatten)]
    sql_type: sql_type::Type,
    sql_column_name: String,
    graphql_field_name: String,
    graphql_type_name: String,
    graphql_order_by_asc: String,
    graphql_order_by_desc: String,
  },
}

// Tests
#[cfg(test)]
mod tests {
  use super::*;
  use similar_asserts::assert_eq;

  #[test]
  fn serialize() {
    let value = ExplicitField::ScalarDatabaseColumn {
      name: "drone".to_owned(),
      nullable: false,
      sql_type: sql_type::Type::Text,
      sql_column_name: "drone_col".to_owned(),
      graphql_field_name: "drone".to_owned(),
      graphql_type_name: "String".to_owned(),
      graphql_order_by_asc: "drone_ASC".to_owned(),
      graphql_order_by_desc: "drone_DESC".to_owned(),
    };

    match serde_json::to_string_pretty(&value) {
      Ok(string) => {
        assert_eq!(
          string,
          r#"{
  "type": "scalarDatabaseColumn",
  "name": "drone",
  "nullable": false,
  "sqlType": "text",
  "sqlColumnName": "drone_col",
  "graphqlFieldName": "drone",
  "graphqlTypeName": "String",
  "graphqlOrderByAsc": "drone_ASC",
  "graphqlOrderByDesc": "drone_DESC"
}"#
        );
      }
      Err(e) => panic!("{}", e),
    }
  }

  #[test]
  fn serialize_other() {
    let value = ExplicitField::ScalarDatabaseColumn {
      name: "drone".to_owned(),
      nullable: false,
      sql_type: sql_type::Type::Other("Yoo".to_owned()),
      sql_column_name: "drone_col".to_owned(),
      graphql_field_name: "drone".to_owned(),
      graphql_type_name: "String".to_owned(),
      graphql_order_by_asc: "drone_ASC".to_owned(),
      graphql_order_by_desc: "drone_DESC".to_owned(),
    };

    match serde_json::to_string_pretty(&value) {
      Ok(string) => {
        assert_eq!(
          string,
          r#"{
  "type": "scalarDatabaseColumn",
  "name": "drone",
  "nullable": false,
  "sqlType": {
    "other": "Yoo"
  },
  "sqlColumnName": "drone_col",
  "graphqlFieldName": "drone",
  "graphqlTypeName": "String",
  "graphqlOrderByAsc": "drone_ASC",
  "graphqlOrderByDesc": "drone_DESC"
}"#
        );
      }
      Err(e) => panic!("{}", e),
    }
  }

  #[test]
  fn deserialize() {
    let data = r#"
      {
        "name": "id",
        "nullable": false,
        "type": "scalarDatabaseColumn",
        "sqlType": "text",
        "sqlColumnName": "id",
        "graphqlFieldName": "id",
        "graphqlTypeName": "String",
        "graphqlOrderByAsc": "idAsc",
        "graphqlOrderByDesc": "idDesc"
      }
      "#;

    match serde_json::from_str(data) {
      Ok(field) => match field {
        ExplicitField::ScalarDatabaseColumn {
          name,
          nullable,
          sql_type,
          sql_column_name,
          graphql_field_name: _,
          graphql_type_name: _,
          graphql_order_by_asc: _,
          graphql_order_by_desc: _,
        } => {
          assert_eq!(name, "id");
          assert_eq!(nullable, false);
          assert_eq!(sql_type, sql_type::Type::Text);
          assert_eq!(sql_column_name, "id");
        }
      },
      Err(e) => panic!("{}", e),
    }
  }

  #[test]
  fn deserialize_other() {
    let data = r#"
      {
        "name": "id",
        "type": "scalarDatabaseColumn",
        "nullable": true,
        "sqlType": {
          "other": "Hii"
        },
        "sqlColumnName": "id",
        "graphqlFieldName": "id",
        "graphqlTypeName": "String",
        "graphqlOrderByAsc": "idAsc",
        "graphqlOrderByDesc": "idDesc"
      }
      "#;

    match serde_json::from_str(data) {
      Ok(field) => match field {
        ExplicitField::ScalarDatabaseColumn {
          name,
          nullable,
          sql_type,
          sql_column_name,
          graphql_field_name: _,
          graphql_type_name: _,
          graphql_order_by_asc: _,
          graphql_order_by_desc: _,
        } => {
          assert_eq!(name, "id");
          assert_eq!(nullable, true);
          assert_eq!(sql_type, sql_type::Type::Other("Hii".to_owned()));
          assert_eq!(sql_column_name, "id");
        }
      },
      Err(e) => panic!("{}", e),
    }
  }
}
