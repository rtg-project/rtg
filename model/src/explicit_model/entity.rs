use super::field::{ExplicitField, ImplicitField};
use partial_struct::PartialStruct;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::rc::Rc;

#[derive(PartialStruct, Serialize, Deserialize, Debug, JsonSchema)]
#[serde(tag = "type", rename_all = "camelCase")]
#[partial(name = "ImplicitEntity")]
#[partial_attribute(derive(Serialize, Deserialize, Debug, JsonSchema))]
#[partial_attribute(serde(tag = "type", rename_all = "camelCase"))]
pub enum ExplicitEntity {
  #[serde(rename_all = "camelCase")]
  DatabaseTable {
    name: String,
    sql_schema_name: String,
    sql_table_name: String,
    graphql_entity_type_name: String,
    graphql_filter_type_name: String,
    graphql_get_single_operation_name: String,
    graphql_get_list_operation_name: String,
    graphql_get_connection_operation_name: String,
    graphql_default_order_by: String,
    graphql_default_first: i16,
    graphql_default_offset: i16,
    #[partial(nested_type = "Vec<Rc<ImplicitField>>")]
    fields: Vec<Rc<ExplicitField>>,
  },
}

// Tests
#[cfg(test)]
mod tests {
  use super::*;
  use crate::explicit_model::sql_type;
  use similar_asserts::assert_eq;

  #[test]
  fn serialize() {
    let value = ExplicitEntity::DatabaseTable {
      name: "person".to_owned(),
      sql_schema_name: "public".to_owned(),
      sql_table_name: "person_table".to_owned(),
      graphql_entity_type_name: "Person".to_owned(),
      graphql_filter_type_name: "PersonWhereFilter".to_owned(),
      graphql_get_single_operation_name: "person".to_owned(),
      graphql_get_list_operation_name: "persons".to_owned(),
      graphql_get_connection_operation_name: "personConnection".to_owned(),
      graphql_default_order_by: "id_ASC".to_owned(),
      graphql_default_first: 10,
      graphql_default_offset: 0,
      fields: vec![
        Rc::new(ExplicitField::ScalarDatabaseColumn {
          name: "id".to_owned(),
          nullable: false,
          sql_type: sql_type::Type::Text,
          sql_column_name: "id_col".to_owned(),
          graphql_field_name: "id".to_owned(),
          graphql_type_name: "String".to_owned(),
          graphql_order_by_asc: "id_ASC".to_owned(),
          graphql_order_by_desc: "id_DESC".to_owned(),
        }),
        Rc::new(ExplicitField::ScalarDatabaseColumn {
          name: "drone".to_owned(),
          nullable: false,
          sql_type: sql_type::Type::Text,
          sql_column_name: "drone_col".to_owned(),
          graphql_field_name: "drone".to_owned(),
          graphql_type_name: "String".to_owned(),
          graphql_order_by_asc: "drone_ASC".to_owned(),
          graphql_order_by_desc: "drone_DESC".to_owned(),
        }),
      ],
    };

    match serde_json::to_string_pretty(&value) {
      Ok(string) => {
        assert_eq!(
          string,
          r#"{
  "type": "databaseTable",
  "name": "person",
  "sqlSchemaName": "public",
  "sqlTableName": "person_table",
  "graphqlEntityTypeName": "Person",
  "graphqlFilterTypeName": "PersonWhereFilter",
  "graphqlGetSingleOperationName": "person",
  "graphqlGetListOperationName": "persons",
  "graphqlGetConnectionOperationName": "personConnection",
  "graphqlDefaultOrderBy": "id_ASC",
  "graphqlDefaultFirst": 10,
  "graphqlDefaultOffset": 0,
  "fields": [
    {
      "type": "scalarDatabaseColumn",
      "name": "id",
      "nullable": false,
      "sqlType": "text",
      "sqlColumnName": "id_col",
      "graphqlFieldName": "id",
      "graphqlTypeName": "String",
      "graphqlOrderByAsc": "id_ASC",
      "graphqlOrderByDesc": "id_DESC"
    },
    {
      "type": "scalarDatabaseColumn",
      "name": "drone",
      "nullable": false,
      "sqlType": "text",
      "sqlColumnName": "drone_col",
      "graphqlFieldName": "drone",
      "graphqlTypeName": "String",
      "graphqlOrderByAsc": "drone_ASC",
      "graphqlOrderByDesc": "drone_DESC"
    }
  ]
}"#
        );
      }
      Err(e) => panic!("{}", e),
    }
  }
}
