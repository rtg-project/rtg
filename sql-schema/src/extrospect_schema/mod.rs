pub mod conversion_error;
pub mod convert_entity;
pub mod convert_field;
pub mod convert_model;

// Tests
#[cfg(test)]
mod tests {

  use super::*;
  use insta::assert_snapshot;
  use rtg_model::explicit_model::{ExplicitEntity, ExplicitField, ExplicitModel, Type};
  use std::rc::Rc;

  #[test]
  fn it_works() {
    let value = ExplicitModel::V1 {
      entities: vec![Rc::new(ExplicitEntity::DatabaseTable {
        name: "person".to_owned(),
        sql_schema_name: "public".to_owned(),
        sql_table_name: "person_table".to_owned(),
        graphql_entity_type_name: "Person".to_owned(),
        graphql_filter_type_name: "PersonWhereFilter".to_owned(),
        graphql_get_single_operation_name: "person".to_owned(),
        graphql_get_list_operation_name: "persons".to_owned(),
        graphql_get_connection_operation_name: "personConnection".to_owned(),
        graphql_default_order_by: "id_ASC".to_owned(),
        graphql_default_first: 10,
        graphql_default_offset: 0,
        fields: vec![
          Rc::new(ExplicitField::ScalarDatabaseColumn {
            name: "id".to_owned(),
            nullable: false,
            sql_type: Type::Text,
            sql_column_name: "id_col".to_owned(),
            graphql_field_name: "id".to_owned(),
            graphql_type_name: "String".to_owned(),
            graphql_order_by_asc: "id_ASC".to_owned(),
            graphql_order_by_desc: "id_DESC".to_owned(),
          }),
          Rc::new(ExplicitField::ScalarDatabaseColumn {
            name: "drone".to_owned(),
            nullable: false,
            sql_type: Type::Text,
            sql_column_name: "drone_col".to_owned(),
            graphql_field_name: "drone".to_owned(),
            graphql_type_name: "String".to_owned(),
            graphql_order_by_asc: "drone_ASC".to_owned(),
            graphql_order_by_desc: "drone_DESC".to_owned(),
          }),
        ],
      })],
    };

    assert_snapshot!(convert_model::convert_model(&value).unwrap());
  }
}
