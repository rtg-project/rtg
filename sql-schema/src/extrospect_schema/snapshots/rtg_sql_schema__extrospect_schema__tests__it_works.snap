---
source: sql-schema/src/extrospect_schema/mod.rs
expression: "convert_model::convert_model(&value).unwrap()"

---
create table "public"."person_table" (
  "id_col" Text,
  "drone_col" Text
);
